Teams/03 Authentication/Demos/01-add-authentication-tab/solution/officedev-talent-management/Tabs/auth-start.html

```diff
@@ -0,0 +1,124 @@
+﻿<!-- Copyright (c) Microsoft Corporation. All rights reserved.
+     Licensed under the MIT License. -->
+<html>
+
+<head>
+  <title>Silent Authentication Sample Login</title>
+</head>
+
+<body>
+  <script src="https://statics.teams.microsoft.com/sdk/v1.2/js/MicrosoftTeams.min.js"
+    integrity="sha384-OncOcMprEYFkUvhe15zi8VdCZIOcuNGnf4ilq2yTfDEjTwPF19V7pZrzxOq/iqt0"
+    crossorigin="anonymous"></script>
+  <script src="https://secure.aadcdn.microsoftonline-p.com/lib/1.0.15/js/adal.min.js"
+    integrity="sha384-lIk8T3uMxKqXQVVfFbiw0K/Nq+kt1P3NtGt/pNexiDby2rKU6xnDY8p16gIwKqgI"
+    crossorigin="anonymous"></script>
+
+  <script type="text/javascript">
+    microsoftTeams.initialize();
+    microsoftTeams.getContext(function (context) {
+      // Generate random state string and store it, so we can verify it in the callback
+      let state = _guid(); // _guid() is a helper function in the sample
+      localStorage.setItem("trainingContent.state", state);
+      localStorage.removeItem("trainingContent.error");
+      // Go to the Azure AD authorization endpoint
+      let queryParams = {
+        client_id: "[AzureAppID]",
+        response_type: "id_token token",
+        response_mode: "fragment",
+        resource: "https://graph.microsoft.com/",
+        redirect_uri: window.location.origin + "/Tabs/auth-end.html",
+        nonce: _guid(),
+        state: state,
+        // The context object is populated by Teams; the loginHint attribute
+        // is used as hinting information
+        login_hint: context.loginHint
+      };
+
+      let authorizeEndpoint = "https://login.microsoftonline.com/" + context.tid + "/oauth2/authorize?" + toQueryString(queryParams);
+      window.location.assign(authorizeEndpoint);
+    });
+
+    // Build query string from map of query parameter
+    function toQueryString(queryParams) {
+      let encodedQueryParams = [];
+      for (let key in queryParams) {
+        encodedQueryParams.push(key + "=" + encodeURIComponent(queryParams[key]));
+      }
+      return encodedQueryParams.join("&");
+    }
+
+    // Generates RFC4122 version 4 guid (128 bits)
+    // (From ADAL.js: https://github.com/AzureAD/azure-activedirectory-library-for-js/blob/dev/lib/adal.js)
+    function _guid() {
```

Comment:
Kind of odd to copy/paste this in, rather than using a library...

Reply:
It's also pretty non-standard to use a non-randomly generated GUID for a nonce isn't it? I thought they were supposed to be as randomly generated as possible?